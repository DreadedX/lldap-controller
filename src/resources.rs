use chrono::{DateTime, Utc};
use kube::CustomResource;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

#[derive(CustomResource, Deserialize, Serialize, Clone, Debug, JsonSchema)]
#[kube(
    kind = "ServiceUser",
    group = "lldap.huizinga.dev",
    version = "v1",
    namespaced,
    status = "ServiceUserStatus"
)]
#[kube(
    shortname = "lsu",
    doc = "Custom resource for managing Service Users inside of LLDAP",
    printcolumn = r#"{"name":"Manager", "type":"boolean", "description":"Can the service user manage passwords", "jsonPath":".spec.passwordManager"}"#,
    printcolumn = r#"{"name":"Age", "type":"date", "jsonPath":".metadata.creationTimestamp"}"#,
    printcolumn = r#"{"name":"Secret", "type":"date", "description":"Secret creation timestamp", "jsonPath":".status.secret_created"}"#
)]
#[serde(rename_all = "camelCase")]
pub struct ServiceUserSpec {
    #[serde(default)]
    password_manager: bool,
    #[serde(default)]
    additional_groups: Vec<String>,
}

#[derive(Deserialize, Serialize, Clone, Default, Debug, JsonSchema)]
pub struct ServiceUserStatus {
    pub secret_created: Option<DateTime<Utc>>,
}

#[cfg(test)]
mod tests {
    use super::*;
    use kube::CustomResourceExt;

    #[test]
    fn service_user_crd_output() {
        insta::assert_yaml_snapshot!(ServiceUser::crd());
    }
}
