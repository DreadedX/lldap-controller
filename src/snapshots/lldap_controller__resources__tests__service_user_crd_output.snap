---
source: src/resources.rs
expression: "ServiceUser::crd()"
---
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  name: serviceusers.lldap.huizinga.dev
spec:
  group: lldap.huizinga.dev
  names:
    categories: []
    kind: ServiceUser
    plural: serviceusers
    shortNames:
      - lsu
    singular: serviceuser
  scope: Namespaced
  versions:
    - additionalPrinterColumns:
        - description: Can the service user manage passwords
          jsonPath: ".spec.passwordManager"
          name: Manager
          type: boolean
        - jsonPath: ".metadata.creationTimestamp"
          name: Age
          type: date
        - description: Secret creation timestamp
          jsonPath: ".status.secret_created"
          name: Secret
          type: date
      name: v1
      schema:
        openAPIV3Schema:
          description: Custom resource for managing Service Users inside of LLDAP
          properties:
            spec:
              properties:
                additionalGroups:
                  default: []
                  items:
                    type: string
                  type: array
                passwordManager:
                  default: false
                  type: boolean
              type: object
            status:
              nullable: true
              properties:
                secret_created:
                  format: date-time
                  nullable: true
                  type: string
              type: object
          required:
            - spec
          title: ServiceUser
          type: object
      served: true
      storage: true
      subresources:
        status: {}
